{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5f00f5e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas_genomics.scalars import Variant\n",
    "from pandas_genomics import sim, io, scalars, GenotypeDtype\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import clarite\n",
    "from clarite import analyze\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9acf2964",
   "metadata": {},
   "outputs": [],
   "source": [
    "start = time.time()\n",
    "\n",
    "# Define the variant for two SNPs\n",
    "\n",
    "variant1 = Variant(\"1\", 1, id=\"rs1\", ref=\"A\", alt=[\"C\"])\n",
    "variant2 = Variant(\"1\", 2, id=\"rs2\", ref=\"G\", alt=[\"T\"])\n",
    "\n",
    "# Define Case-Control ratio\n",
    "num_samples = 1000\n",
    "case_control_ratio = \"1:3\"\n",
    "n_cases = int(num_samples / 4)\n",
    "n_controls = num_samples - n_cases\n",
    "PEN_BASE = 0.05\n",
    "PEN_DIFF = 0.25\n",
    "MAFA = 0.05\n",
    "MAFB = 0.05\n",
    "SNR = 0.01\n",
    "# Result files\n",
    "EDGE_alpha_Final = pd.DataFrame()\n",
    "Recessive_Results_Final = pd.DataFrame()\n",
    "Sub_Additive_Results_Final = pd.DataFrame()\n",
    "Additive_Results_Final = pd.DataFrame()\n",
    "Super_Additive_Results_Final = pd.DataFrame()\n",
    "Dominant_Results_Final = pd.DataFrame()\n",
    "Het_Results_Final = pd.DataFrame()\n",
    "EDGE_Results_Final = pd.DataFrame()\n",
    "NULL_Results_Final = pd.DataFrame()\n",
    "All_Results_Final = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4c7fe427",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas_genomics.sim.biallelic_model_simulator.BAMS at 0x7ff389693460>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_seed = 1\n",
    "test_seed = 2001\n",
    "\n",
    "startcycle = time.time()\n",
    "# Recessive Main Effect for SNP1 without interaction\n",
    "# Training data\n",
    "eff1 = sim.SNPEffectEncodings.RECESSIVE\n",
    "eff2 = sim.SNPEffectEncodings.ADDITIVE\n",
    "\n",
    "train_rec_main_effect00 = sim.BAMS.from_model(\n",
    "        eff1=sim.SNPEffectEncodings.RECESSIVE,\n",
    "        eff2=sim.SNPEffectEncodings.ADDITIVE,\n",
    "        penetrance_base=PEN_BASE,\n",
    "        penetrance_diff=PEN_DIFF,\n",
    "        main1=1,\n",
    "        main2=0,\n",
    "        interaction=0,\n",
    "        snp1=variant1,\n",
    "        snp2=variant2,\n",
    "        random_seed=train_seed,\n",
    "    )\n",
    "\n",
    "train_rec_main_effect00   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4c17bdd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>SNP1</th>\n",
       "      <th>SNP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/C</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Case</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>Case</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/C</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>Case</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>Case</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Outcome SNP1 SNP2\n",
       "0    Control  A/A  G/G\n",
       "1    Control  A/C  G/G\n",
       "2    Control  A/A  G/G\n",
       "3    Control  A/A  G/G\n",
       "4       Case  A/A  G/G\n",
       "..       ...  ...  ...\n",
       "995  Control  A/A  G/G\n",
       "996     Case  A/A  G/G\n",
       "997  Control  A/C  G/G\n",
       "998     Case  A/A  G/G\n",
       "999     Case  A/A  G/G\n",
       "\n",
       "[1000 rows x 3 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_rec_me_pb000 = train_rec_main_effect00.generate_case_control(\n",
    "    n_cases=n_cases, n_controls=n_controls, maf1=MAFA, maf2=MAFB, snr=SNR)\n",
    "\n",
    "train_rec_me_pb000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "012a364a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Binary Outcome (family = Binomial): 'Outcome'\n",
      "\t250 occurrences of 'Case' coded as 0\n",
      "\t750 occurrences of 'Control' coded as 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TestSeed</th>\n",
       "      <th>TrainSeed</th>\n",
       "      <th>BioAct</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>Alpha Value</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>Minor Allele Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2001</td>\n",
       "      <td>1</td>\n",
       "      <td>Recessive</td>\n",
       "      <td>rs1</td>\n",
       "      <td>-1.844349</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2001</td>\n",
       "      <td>1</td>\n",
       "      <td>Recessive</td>\n",
       "      <td>rs2</td>\n",
       "      <td>0.050942</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   TestSeed  TrainSeed     BioAct Variant ID  Alpha Value Ref Allele  \\\n",
       "0      2001          1  Recessive        rs1    -1.844349          A   \n",
       "1      2001          1  Recessive        rs2     0.050942          G   \n",
       "\n",
       "  Alt Allele  Minor Allele Frequency  \n",
       "0          C                  0.0535  \n",
       "1          T                  0.0540  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_weights_rec_me_pb000 = (train_rec_me_pb000.genomics.calculate_edge_encoding_values(\n",
    "            data=train_rec_me_pb000[\"Outcome\"], outcome_variable=\"Outcome\"\n",
    "        )\n",
    ")\n",
    "edge_weights_rec_me = edge_weights_rec_me_pb000.copy()\n",
    "edge_weights_rec_me.insert(loc=0, column=\"BioAct\", value=\"Recessive\")\n",
    "edge_weights_rec_me.insert(loc=0, column=\"TrainSeed\", value=train_seed)\n",
    "edge_weights_rec_me.insert(loc=0, column=\"TestSeed\", value=test_seed)\n",
    "edge_weights_rec_me"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "fa738755",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test data\n",
    "test_rec_main_effect00 = sim.BAMS.from_model(\n",
    "        eff1=sim.SNPEffectEncodings.RECESSIVE,\n",
    "        eff2=sim.SNPEffectEncodings.ADDITIVE,\n",
    "        penetrance_base=PEN_BASE,\n",
    "        penetrance_diff=PEN_DIFF,\n",
    "        main1=1,\n",
    "        main2=0,\n",
    "        interaction=0,\n",
    "        snp1=variant1,\n",
    "        snp2=variant2,\n",
    "        random_seed=test_seed,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2ea3d02f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>SNP1</th>\n",
       "      <th>SNP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/C</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>Case</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>Control</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Outcome SNP1 SNP2\n",
       "0    Control  A/C  G/G\n",
       "1    Control  A/A  G/G\n",
       "2    Control  A/A  G/G\n",
       "3    Control  A/A  G/G\n",
       "4    Control  A/A  G/G\n",
       "..       ...  ...  ...\n",
       "995  Control  A/A  G/G\n",
       "996     Case  A/A  G/G\n",
       "997  Control  A/A  G/G\n",
       "998  Control  A/A  G/G\n",
       "999  Control  A/A  G/G\n",
       "\n",
       "[1000 rows x 3 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_rec_me_pb000 = test_rec_main_effect00.generate_case_control(\n",
    "        n_cases=n_cases, n_controls=n_controls, maf1=MAFA, maf2=MAFB, snr=SNR\n",
    "    )\n",
    "#test_rec_me_pb000[\"Outcome\"].cat.reorder_categories([\"Control\", \"Case\"], inplace=True)\n",
    "test_rec_me_pb000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "680850e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Outcome</th>\n",
       "      <th>SNP1</th>\n",
       "      <th>SNP2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>Case</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>Control</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Outcome  SNP1  SNP2\n",
       "0    Control   0.0   0.0\n",
       "1    Control   0.0   0.0\n",
       "2    Control   0.0   0.0\n",
       "3    Control   0.0   0.0\n",
       "4    Control   0.0   0.0\n",
       "..       ...   ...   ...\n",
       "995  Control   0.0   0.0\n",
       "996     Case   0.0   0.0\n",
       "997  Control   0.0   0.0\n",
       "998  Control   0.0   0.0\n",
       "999  Control   0.0   0.0\n",
       "\n",
       "[1000 rows x 3 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run Regression by using weightes from CLARITE\n",
    "# Addtive Encoding\n",
    "test_rec_me_pb000_ADD = test_rec_me_pb000.genomics.encode_additive()\n",
    "test_rec_me_pb000_ADD\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "be56266e",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'clarite.modules.analyze' has no attribute 'association_study'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/tv/qf2nl0h10kn_55zkv8p1wy3r0000gn/T/ipykernel_2673/3164041359.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0madd_results_rec_me_pb000\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0manalyze\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0massociation_study\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtest_rec_me_pb000_ADD\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutcomes\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Outcome\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0madd_results_rec_me_pb000\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"odds ratio\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madd_results_rec_me_pb000\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Beta\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0madd_results_rec_me_pb000\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Encoding\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Additive\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0madd_results_rec_me_pb000\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"BioAct\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Recessive\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0madd_results_rec_me_pb000\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"TrainSeed\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtrain_seed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'clarite.modules.analyze' has no attribute 'association_study'"
     ]
    }
   ],
   "source": [
    "add_results_rec_me_pb000 = analyze.association_study(data=test_rec_me_pb000_ADD, outcomes=\"Outcome\")\n",
    "add_results_rec_me_pb000[\"odds ratio\"] = np.exp(add_results_rec_me_pb000[\"Beta\"])\n",
    "add_results_rec_me_pb000.insert(loc=0, column=\"Encoding\", value=\"Additive\")\n",
    "add_results_rec_me_pb000.insert(loc=0, column=\"BioAct\", value=\"Recessive\")\n",
    "add_results_rec_me_pb000.insert(loc=0, column=\"TrainSeed\", value=train_seed)\n",
    "add_results_rec_me_pb000.insert(loc=0, column=\"TestSeed\", value=test_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cd8a518",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
